File: rust/cumulus-dotfiles/src/sysutils.rs
//! Session utilities that wrap Sway/Wayland tools: the themed swaylock wrapper
//! (`lock.sh`), the swayidle daemon (`idle.sh`), and the grim/slurp screenshot
//! helper (`screenshot.sh`).

use crate::context::Context;
use crate::error::{Error, Result};
use crate::theme::palette::Palette;
use crate::theme::state::State;
use crate::util;
use std::io::Write;
use std::os::unix::process::CommandExt;
use std::process::{Command, Stdio};

fn strip_hash(s: &str) -> String {
    s.strip_prefix('#').unwrap_or(s).to_string()
}

/// `cumulus-lock` — swaylock styled to follow the active theme. Execs swaylock.
pub fn run_lock(ctx: &Context) -> Result<()> {
    let state = State::load(ctx);
    let mut flavor = state.flavor;
    if flavor.is_empty() {
        flavor = "oci".to_string();
    }
    let palette_path = ctx.palettes_dir().join(format!("{flavor}.sh"));
    let palette = if palette_path.is_file() {
        Palette::load(ctx, &flavor)
    } else {
        Palette::load(ctx, "oci")
    };
    let base = strip_hash(palette.get("BASE"));
    let blue = strip_hash(palette.get("BLUE"));
    let text = strip_hash(palette.get("TEXT"));
    let green = strip_hash(palette.get("GREEN"));

    let err = Command::new("swaylock")
        .args([
            "--color",
            &base,
            "--inside-color",
            &base,
            "--ring-color",
            &blue,
            "--line-color",
            &base,
            "--text-color",
            &text,
            "--inside-ver-color",
            &blue,
            "--ring-ver-color",
            &blue,
            "--key-hl-color",
            &green,
            "--separator-color",
            &base,
            "--font",
            "JetBrainsMono Nerd Font",
            "--indicator-radius",
            "100",
            "--indicator-thickness",
            "10",
        ])
        .exec();
    Err(Error::new(format!("failed to exec swaylock: {err}")))
}

/// Resolve a sibling `cumulus-<name>` binary next to the running executable so
/// the daemon can invoke it whether installed or run from the build tree.
fn sibling_command(name: &str) -> String {
    if let Ok(exe) = std::env::current_exe() {
        if let Some(dir) = exe.parent() {
            let candidate = dir.join(name);
            if candidate.exists() {
                return candidate.to_string_lossy().into_owned();
            }
        }
    }
    name.to_string()
}

/// `cumulus-idle` — swayidle daemon (auto-lock, dpms, suspend). Execs swayidle.
pub fn run_idle(_ctx: &Context) -> Result<()> {
    let lock_cmd = sibling_command("cumulus-lock");
    let err = Command::new("swayidle")
        .args([
            "-w",
            "timeout",
            "300",
            &lock_cmd,
            "timeout",
            "600",
            "swaymsg \"output * dpms off\"",
            "resume",
            "swaymsg \"output * dpms on\"",
            "timeout",
            "900",
            "systemctl suspend",
            "before-sleep",
            &lock_cmd,
        ])
        .exec();
    Err(Error::new(format!("failed to exec swayidle: {err}")))
}

fn notify(msg: &str) {
    if util::command_exists("notify-send") {
        let _ = Command::new("notify-send")
            .args(["Screenshot", msg])
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
    }
}

/// Capture the focused window geometry via `swaymsg -t get_tree | jq ...`.
fn focused_window_geometry() -> Option<String> {
    let tree = Command::new("swaymsg")
        .args(["-t", "get_tree"])
        .stderr(Stdio::null())
        .output()
        .ok()?;
    if !tree.status.success() {
        return None;
    }
    let mut jq = Command::new("jq")
        .args([
            "-r",
            r#".. | select(.focused? == true) | .rect | "\(.x),\(.y) \(.width)x\(.height)""#,
        ])
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .spawn()
        .ok()?;
    jq.stdin.take()?.write_all(&tree.stdout).ok()?;
    let out = jq.wait_with_output().ok()?;
    let geom = String::from_utf8_lossy(&out.stdout).trim().to_string();
    (!geom.is_empty()).then_some(geom)
}

/// `cumulus-screenshot [full|region|window]` — grim/slurp screenshot helper.
pub fn run_screenshot(ctx: &Context, args: &[String]) -> Result<()> {
    let mode = args.first().map(String::as_str).unwrap_or("region");
    let dir = ctx.home.join("Pictures/Screenshots");
    std::fs::create_dir_all(&dir)?;
    let stamp = timestamp();
    let file = dir.join(format!("{stamp}.png"));

    match mode {
        "full" => {
            grim(&[file.to_string_lossy().as_ref()])?;
        }
        "region" => {
            let geom = Command::new("slurp").output();
            let geom = match geom {
                Ok(o) if o.status.success() => {
                    String::from_utf8_lossy(&o.stdout).trim().to_string()
                }
                _ => {
                    notify("Cancelled");
                    return Ok(());
                }
            };
            grim(&["-g", &geom, file.to_string_lossy().as_ref()])?;
        }
        "window" => {
            let Some(geom) = focused_window_geometry() else {
                notify("No focused window found");
                return Err(Error::with_code("", 1));
            };
            grim(&["-g", &geom, file.to_string_lossy().as_ref()])?;
        }
        _ => {
            eprintln!("Usage: cumulus-screenshot {{full|region|window}}");
            return Err(Error::with_code("", 1));
        }
    }

    // Copy the saved image to the clipboard, then notify.
    if let Ok(bytes) = std::fs::read(&file) {
        if let Ok(mut child) = Command::new("wl-copy")
            .stdin(Stdio::piped())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()
        {
            if let Some(mut stdin) = child.stdin.take() {
                let _ = stdin.write_all(&bytes);
            }
            let _ = child.wait();
        }
    }
    notify(&format!(
        "Saved to {} (copied to clipboard)",
        file.display()
    ));
    Ok(())
}

fn grim(args: &[&str]) -> Result<()> {
    let status = Command::new("grim").args(args).status()?;
    if status.success() {
        Ok(())
    } else {
        Err(Error::with_code("grim failed", 1))
    }
}

/// `date +%Y-%m-%d_%H-%M-%S` in local time, via the `date` command for parity.
fn timestamp() -> String {
    Command::new("date")
        .arg("+%Y-%m-%d_%H-%M-%S")
        .output()
        .ok()
        .filter(|o| o.status.success())
        .map(|o| String::from_utf8_lossy(&o.stdout).trim().to_string())
        .unwrap_or_default()
}

File: src/main/scala/cumulus/dotfiles/sysutils/SysUtils.scala
package cumulus.dotfiles.sysutils

import cumulus.dotfiles.context.Context
import cumulus.dotfiles.error.{CommandError, CumulusError}
import cumulus.dotfiles.theme.ThemeEngine

object SysUtils:
  def runLock(ctx: Context): Either[CumulusError, Unit] =
    println("\u001b[1;34m[cumulus lock]\u001b[0m Locking screen via swaylock...")
    val lockConfigFile = ctx.configDir / "swaylock" / "config"
    try
      val cmd = if os.exists(lockConfigFile) then
        Seq("swaylock", "-f", "--config", lockConfigFile.toString)
      else
        val palette = ThemeEngine.getActivePalette(ctx)
        val baseHex = palette.base.stripPrefix("#")
        Seq("swaylock", "-f", "-c", baseHex)
      os.proc(cmd).call(check = false)
      Right(())
    catch
      case e: Exception => Left(CommandError(s"Lock failed: ${e.getMessage}"))

  def runIdle(ctx: Context): Either[CumulusError, Unit] =
    println("\u001b[1;34m[cumulus idle]\u001b[0m Launching swayidle daemon...")
    val lockConfigFile = ctx.configDir / "swaylock" / "config"
    val lockCmd = if os.exists(lockConfigFile) then
      s"swaylock -f --config $lockConfigFile"
    else
      "swaylock -f -c 1e1e2e"

    try
      val res = os.proc(
        "swayidle", "-w",
        "timeout", "300", lockCmd,
        "timeout", "600", "swaymsg 'output * dpms off'",
        "resume", "swaymsg 'output * dpms on'",
        "timeout", "900", "systemctl suspend",
        "before-sleep", lockCmd
      ).call(check = false)
      Right(())
    catch
      case e: Exception => Left(CommandError(s"Idle daemon failed: ${e.getMessage}"))

  def runScreenshot(ctx: Context, args: List[String]): Either[CumulusError, Unit] =
    val mode = args.headOption.getOrElse("region")

    if mode != "full" && mode != "region" && mode != "window" then
      System.err.println("Usage: cumulus-screenshot {full|region|window}")
      return Left(CommandError("Usage: cumulus-screenshot {full|region|window}", 1))

    val screenshotsDir = ctx.home / "Pictures" / "Screenshots"
    os.makeDir.all(screenshotsDir)
    val timestamp = os.proc("date", "+%Y-%m-%d_%H-%M-%S").call(check = false).out.text().trim
    val file = screenshotsDir / s"$timestamp.png"

    println(s"\u001b[1;34m[cumulus screenshot]\u001b[0m Capturing $mode screenshot to $file...")

    try
      val captureRes = mode match
        case "full" =>
          os.proc("grim", file.toString).call(check = false)
        case "region" =>
          val slurpRes = os.proc("slurp").call(
            check = false,
            stdin = os.Inherit,
            stdout = os.Pipe,
            stderr = os.Inherit
          )
          if slurpRes.exitCode == 0 && slurpRes.out.text().trim.nonEmpty then
            val geom = slurpRes.out.text().trim
            os.proc("grim", "-g", geom, file.toString).call(check = false)
          else
            notifyDesktop("Cancelled", "Screenshot region selection cancelled.")
            return Right(())
        case "window" =>
          getFocusedWindowGeometry() match
            case Some(geom) =>
              os.proc("grim", "-g", geom, file.toString).call(check = false)
            case None =>
              notifyDesktop("Failed", "No focused window found.")
              return Left(CommandError("No focused window found", 1))

      if captureRes.exitCode == 0 && os.exists(file) then
        // Copy screenshot image bytes to wl-copy clipboard if available
        try
          if isCommandAvailable("wl-copy") then
            os.proc("wl-copy").call(stdin = os.read.bytes(file), check = false)
        catch
          case _: Exception => ()

        notifyDesktop("Screenshot Saved", s"Saved to ${file.toString} (copied to clipboard)")
        println(s"  \u001b[32m[OK]\u001b[0m Saved to $file (copied to clipboard)")
        Right(())
      else
        Left(CommandError("grim screenshot capture failed", 1))
    catch
      case e: Exception => Left(CommandError(s"Screenshot failed: ${e.getMessage}"))

  private def getFocusedWindowGeometry(): Option[String] =
    try
      val res = os.proc("swaymsg", "-t", "get_tree").call(check = false)
      if res.exitCode == 0 then
        val json = ujson.read(res.out.text())
        findFocusedNodeGeometry(json)
      else None
    catch
      case _: Exception => None

  private def findFocusedNodeGeometry(node: ujson.Value): Option[String] =
    try
      if node.obj.get("focused").exists(_.bool) then
        val rect = node.obj("rect")
        val x = rect("x").num.toInt
        val y = rect("y").num.toInt
        val w = rect("width").num.toInt
        val h = rect("height").num.toInt
        Some(s"$x,$y ${w}x$h")
      else
        val tiledNodes = node.obj.get("nodes").map(_.arr).getOrElse(Vector.empty)
        val floatingNodes = node.obj.get("floating_nodes").map(_.arr).getOrElse(Vector.empty)
        (tiledNodes ++ floatingNodes).flatMap(findFocusedNodeGeometry).headOption
    catch
      case _: Exception => None

  private def notifyDesktop(title: String, body: String): Unit =
    try
      if isCommandAvailable("notify-send") then
        val escapedTitle = title.replace("\\", "\\\\").replace("\"", "\\\"")
        val escapedBody = body.replace("\\", "\\\\").replace("\"", "\\\"")
        os.proc("notify-send", "-u", "normal", "-t", "4000", "-a", "cumulus", escapedTitle, escapedBody).call(check = false)
    catch
      case _: Exception => () // Silently fail if notification daemon unavailable

  def runCalendar(ctx: Context): Either[CumulusError, Unit] =
    if isCommandAvailable("swaync-client") && isProcessRunning("swaync") then
      try
        os.proc("swaync-client", "-t", "-sw").call(check = false)
        Right(())
      catch
        case _: Exception => runFallbackCalendar(ctx)
    else
      runFallbackCalendar(ctx)

  private def isProcessRunning(name: String): Boolean =
    try os.proc("pgrep", "-x", name).call(check = false).exitCode == 0 catch case _: Exception => false

  private def runFallbackCalendar(ctx: Context): Either[CumulusError, Unit] =
    try
      if isCommandAvailable("kitty") then
        os.proc(
          "kitty",
          "--class=cumulus-calendar",
          "--title=Calendar",
          "-o", "font_size=15",
          "-o", "remember_window_size=no",
          "-o", "initial_window_width=680",
          "-o", "initial_window_height=440",
          "sh", "-c", "cal -3; echo ''; read -n 1 -s -r -p '  [Press any key or Escape to close]' || true"
        ).spawn()
        Right(())
      else
        println("  \u001b[33m[NOTE]\u001b[0m Calendar tool not available.")
        Right(())
    catch
      case e: Exception => Left(CommandError(s"Calendar popup failed: ${e.getMessage}"))

  private def isCommandAvailable(cmd: String): Boolean =
    try os.proc("which", cmd).call(check = false).exitCode == 0 catch case _: Exception => false
